e2e/testcases/notification_test.go

```diff
@@ -0,0 +1,173 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package e2e
+
+import (
+	"encoding/base64"
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+	"kpt.dev/configsync/e2e/nomostest"
+	"kpt.dev/configsync/e2e/nomostest/ntopts"
+	"kpt.dev/configsync/e2e/nomostest/taskgroup"
+	nomostesting "kpt.dev/configsync/e2e/nomostest/testing"
+	"kpt.dev/configsync/pkg/api/configsync"
+	"kpt.dev/configsync/pkg/core"
+	"kpt.dev/configsync/pkg/kinds"
+	"kpt.dev/configsync/pkg/reconcilermanager"
+)
+
+func TestNotification(t *testing.T) {
+	rootSyncNN := nomostest.RootSyncNN(configsync.RootSyncName)
+	repoSyncNN := nomostest.RepoSyncNN(backendNamespace, configsync.RepoSyncName)
+	rootReconcilerName := core.RootReconcilerName(rootSyncNN.Name)
+	backendReconcilerName := core.NsReconcilerName(repoSyncNN.Namespace, repoSyncNN.Name)
+	nt := nomostest.New(t, nomostesting.SyncSource,
+		ntopts.InstallNotificationServer,
+		ntopts.NamespaceRepo(repoSyncNN.Namespace, repoSyncNN.Name),
+	)
+	var err error
+	credentialMap := map[string]string{
+		rootSyncNN.Namespace: "pass123",
+		repoSyncNN.Namespace: "pass456",
+	}
+	for ns, pass := range credentialMap {
+		_, err = nomostest.NotificationSecret(nt, ns,
+			nomostest.WithNotificationUsername("user"),
+			nomostest.WithNotificationPassword(pass),
+		)
+		if err != nil {
+			nt.T.Fatal(err)
+		}
+		_, err = nomostest.NotificationConfigMap(nt, ns,
+			nomostest.WithOnSyncSyncedTrigger,
+			nomostest.WithSyncSyncedTemplate,
+			nomostest.WithLocalWebhookService,
+		)
+		if err != nil {
+			nt.T.Fatal(err)
+		}
+	}
+
+	repoSyncBackend := nomostest.RepoSyncObjectV1Beta1FromNonRootRepo(nt, repoSyncNN)
+	rootSync := nomostest.RootSyncObjectV1Beta1FromRootRepo(nt, rootSyncNN.Name)
+	if err := nt.Get(rootSyncNN.Name, rootSyncNN.Namespace, rootSync); err != nil {
+		nt.T.Fatal(err)
+	}
+
+	nomostest.SubscribeRootSyncNotification(rootSync, "on-sync-synced", "local")
+	if err := nt.Update(rootSync); err != nil {
+		nt.T.Fatal(err)
+	}
+	err = nomostest.WatchObject(nt, kinds.Deployment(), rootReconcilerName, rootSyncNN.Namespace,
+		[]nomostest.Predicate{nomostest.DeploymentHasContainer(reconcilermanager.Notification)})
+	if err != nil {
+		nt.T.Fatal(err)
+	}
+	records, err := waitForNotifications(nt, 1)
```

Comment:
It may be also worthwhile to validate the `notified` annotation.

Reply:
Sure, I can add validation for that as well

Reply:
done